src/compiler/scala/tools/nsc/typechecker/Typers.scala

```diff
@@ -1091,12 +1091,15 @@ trait Typers extends Adaptations with Tags with TypersTracking with PatternTyper
           @inline def tpdPos(transformed: Tree) = typedPos(tree.pos, mode, pt)(transformed)
           @inline def tpd(transformed: Tree)    = typed(transformed, mode, pt)
 
-          @inline def warnValueDiscard(): Unit = if (!isPastTyper && settings.warnValueDiscard) {
+          @inline def warnValueDiscard(): Unit = if (!isPastTyper) {
             def isThisTypeResult = (tree, tree.tpe) match {
               case (Apply(Select(receiver, _), _), SingleType(_, sym)) => sym == receiver.symbol
               case _ => false
             }
-            if (!isThisTypeResult) context.warning(tree.pos, "discarded non-Unit value")
+            if (tree.tpe.termSymbol == UnitClass.companionModule)
+              context.warning(tree.pos, "discarded companion object Unit of type Unit.type. Use () to obtain a value of type Unit.")
```

Comment:
Or pedantically, `Use '()' for the value of type Unit.` There is only one value, and I could imagine misunderstanding the current text to mean I should write `Unit()` to produce a value. The word `type` might sound ambiguous, that it's the companion of type Unit.type. `discarded Unit, the companion object of class Unit. Use '()' for the unit value, the only value of type Unit.` or `the only value of class Unit`. WDYT? Nice to see this go from forum to PR.

Reply:
Yeah, removing the "of type Unit.type" part definitely makes sense. After looking at both wordings, I think the intent would be clearer if we just avoided the "use" word:
```
discarded Unit, the companion object of class Unit. The only value of type Unit is `()`
```
Do you think it's better that way?

Reply:
Error messages are definitely a harder problem than naming in comp sci. I prefer when it tells me what to do. There is a condescending message about "Did you mean to write...?" but I don't recommend it. `The unit value is written '()'.` Or `Write '()' for the unit value.` `unit value` is the name for it; I guess it's excessive to explain what it means.